File: src/modes/RunProfile.js
// RunProfile.js
// What a single run *is*: how long it lasts, how hard it gets, which obstacles
// may spawn and when the helper pickups start. One profile is built per run and
// hangs off `game.profile`; the managers read it instead of reaching for
// `game.score` and the global config, which is what makes a second play mode
// possible at all.
// Changes:
// - Open Space weather (pair/combo/focus by KM) lives on OpenWorldProfile.
// - Journey L6+ pairing / L20+ comboTheme / encounters live on JourneyProfile.
// - OPEN_WORLD_UNLOCKS messages are null — types still unlock by KM, but
//   Open Space no longer flashes hazard-name banners.
// - OPEN_WORLD_UNLOCKS: driftCurrent@3500, phase@4500, repulsor@5500,
//   sweepGate@7000 (between/after existing shooting→BH ladder).
// - `focusChance` (default 0.5): how often `focusType` wins a set-piece slot.
// - PLAY_MODE.hazardLab for the Journey-map sandbox.
// - Default `introBeats` wraps `introMessage` as a one-element array.
// - `wallBoostsFromScore` (default 12000): deep-run wall-boost slab gate.
// - `obstaclesFromScore` (default 0): Journey overrides to a shared early KM
//   mark; Open Space spawns as soon as intro ends.
// - Open Space on iOS draw LOD: soft maxOnScreen (18) so late-run fields
//   cannot explode Safari draw cost; Android/desktop stay uncapped.
// - Created file: RunProfile (the contract + shared defaults) and
//   OpenWorldProfile, which reproduces the endless run's existing numbers
//   exactly. The obstacle unlock schedule moved here from ObstacleManager, so
//   there is now one source of truth for it.

import { clamp01 } from '../utils/math.js';
import {
    OPEN_SPACE_PAIRED_FROM_KM,
    weatherAt,
} from '../config/OpenSpaceWeather.js';

export const PLAY_MODE = {
    openWorld: 'openWorld',
    journey: 'journey',
    hazardLab: 'hazardLab',
};

export const PLAY_MODE_STORAGE_KEY = 'playMode';

// Every obstacle type in spawn order, with the Open World distance that unlocks
// it. `message` stays null so ObstacleManager unlocks silently — no hazard
// banners mid-run. ObstacleManager used to keep its own copy of this table.
export const OPEN_WORLD_UNLOCKS = [
    { type: 'simple', score: 0, message: null },
    { type: 'sideBarrier', score: 1000, message: null },
    { type: 'complex', score: 1000, message: null },
    { type: 'moving', score: 2000, message: null },
    { type: 'shooting', score: 3000, message: null },
    { type: 'driftCurrent', score: 3500, message: null },
    { type: 'pulsating', score: 4000, message: null },
    { type: 'phase', score: 4500, message: null },
    { type: 'wormhole', score: 5000, message: null },
    { type: 'repulsor', score: 5500, message: null },
    { type: 'blackhole', score: 6000, message: null },
    { type: 'sweepGate', score: 7000, message: null },
];

export const ALL_OBSTACLE_TYPES = OPEN_WORLD_UNLOCKS.map((entry) => entry.type);

/**
 * The contract the managers rely on. Defaults here describe the endless run, so
 * a subclass only overrides what it actually changes.
 */
export class RunProfile {
    constructor(game) {
        this.game = game;
    }

    // --- Identity --------------------------------------------------------
    get mode() {
        return PLAY_MODE.openWorld;
    }

    /** True when there is no meaningful finish line to show progress against. */
    get isEndless() {
        return true;
    }

    /** Journey level number, or null outside Journey. */
    get level() {
        return null;
    }

    get title() {
        return 'Open Space';
    }

    /** Score at which the run is won. */
    get goalScore() {
        return Infinity;
    }

    /** Journey smash-star target, or null outside Journey. */
    get smashTarget() {
        return null;
    }

    /** A line for the milestone log as the run opens, or null for none. */
    get introMessage() {
        return null;
    }

    /** On-screen intro beats; default is a single introMessage line if any. */
    get introBeats() {
        const line = this.introMessage;
        return line ? [line] : null;
    }

    /** Whether finishing this run offers the leaderboard. */
    get submitsScore() {
        return true;
    }

    // --- Flight ----------------------------------------------------------
    get runsTutorial() {
        return true;
    }

    /** Scales the ship's forward speed, and so the whole run's pace. */
    get speedMultiplier() {
        return 1;
    }

    // --- Pickups ---------------------------------------------------------
    get shieldsFromScore() {
        return 500;
    }

    /** Deep-run wall-boost slabs; Open World / long flights only by default. */
    get wallBoostsFromScore() {
        return 12000;
    }

    get collectiblesFromScore() {
        return 100;
    }

    // --- Obstacles -------------------------------------------------------
    /** HUD KM before obstacle rows may spawn. 0 = as soon as intro unpauses. */
    get obstaclesFromScore() {
        return 0;
    }

    /**
     * How many obstacles may be waiting ahead of the ship at once. Open World
     * has never had an effective cap: the old `length < 7` test guarded a spawn
     * branch that can only fire once per run, so a row was never actually
     * withheld. Keeping it uncapped is what "plays identically" means here.
     */
    get maxOnScreen() {
        return Infinity;
    }

    /** Chance a spawn slot takes a plain asteroid cluster rather than a set piece. */
    get simpleChance() {
        return 0.65;
    }

    /** A type this run leans on, giving otherwise-equal levels their own feel. */
    get focusType() {
        return null;
    }

    /** Chance a set-piece slot picks `focusType` when one is set (0–1). */
    get focusChance() {
        return 0.5;
    }

    /** When true, adjacent row slots may repeat the same set-piece type. */
    get allowAdjacentSetPieces() {
        return false;
    }

    get pairTheme() {
        return null;
    }

    get comboTheme() {
        return null;
    }

    get encounterCount() {
        return 0;
    }

    get isLateJourney() {
        return false;
    }

    get usesPairedBelt() {
        return false;
    }

    rollRowSpawnCount() {
        const maxSpawns = Math.max(1, this.maxRowSpawns());
        if (maxSpawns <= 1) return 1;
        let spawnCount = 1;
        if (maxSpawns >= 2 && Math.random() >= 0.7) spawnCount = 2;
        if (maxSpawns >= 3 && Math.random() >= 0.9) spawnCount = 3;
        return Math.min(spawnCount, maxSpawns);
    }

    get advancedBlackHoles() {
        return this.game.score > 1000;
    }

    /** Vertical spacing between spawn rows, in world units. */
    gapRange(canvasHeight) {
        return { min: canvasHeight * 0.25, max: canvasHeight * 0.4 };
    }

    /** Multiplier feeding the size of plain-asteroid clusters. */
    density() {
        const scaling = this.game.config.obstacles.scaling;
        const progress = Math.min(this.game.score / (scaling.rampUpDistance * 1.2), 1);
        return scaling.startDensity
            + (scaling.maxDensity - scaling.startDensity) * Math.pow(progress, 1.2);
    }

    /** Floor for how many asteroids a cluster contains. */
    baseClusterCount() {
        return 2 + Math.floor(this.game.score / 8000);
    }

    /** Ceiling for a single simple-asteroid cluster. */
    maxClusterCount() {
        return 4;
    }

    /** How many independent obstacles a spawn row may place side by side. */
    maxRowSpawns() {
        return 3;
    }

    /**
     * Types playable at `score`, each with the message to announce it (null =
     * announce nothing). Called every frame, so it must stay cheap.
     */
    unlocksBy(score) {
        return OPEN_WORLD_UNLOCKS.filter((entry) => score >= entry.score);
    }

    // --- Progress --------------------------------------------------------
    isRunComplete(score) {
        return score >= this.goalScore;
    }

    /** 0 - 1 through the run, for the HUD goal bar. */
    progress(score) {
        return Number.isFinite(this.goalScore) ? clamp01(score / this.goalScore) : 0;
    }
}

/**
 * The endless run exactly as it has always played: the ramp in
 * `GameConfig.obstacles.scaling`, unlocks by distance, no finish line in
 * practice. `TOTAL_DISTANCE` is read in the same units the old
 * `remainingDistance` maths used, so the (unreachable) victory line is unmoved.
 */
export class OpenWorldProfile extends RunProfile {
    get goalScore() {
        return this.game.TOTAL_DISTANCE * 100;
    }

    get usesPairedBelt() {
        return this.game.score >= OPEN_SPACE_PAIRED_FROM_KM;
    }

    get pairTheme() {
        return weatherAt(this.game.score).pair;
    }

    get comboTheme() {
        return weatherAt(this.game.score).combo;
    }

    get focusType() {
        return weatherAt(this.game.score).focus;
    }

    /**
     * Uncapped on Android/desktop (historical Open World feel). On iOS Safari
     * a soft ceiling keeps late-run Canvas2D draw+collision from melting FPS;
     * Journey already caps via JourneyProfile.
     */
    get maxOnScreen() {
        return this.game.iosDrawLod ? 18 : Infinity;
    }
}

File: ios-native/SpaceSwoosh/Core/RunProfile.swift
// RunProfile.swift
// Changes: Journey/Lab clusterCount adds Int(roll * dens) like Android
// spawnSimpleAsteroids. Hazard Lab uses advanced black-hole Y-pull (matches
// JS sandbox). Journey L6+ mixed rows / L20+ late knobs; pairTheme +
// comboTheme + encounterCount from spec. Open Space weather, belt density
// (tighter vertical pack by 5k / 12.5k / 20k), and +10% cruise are live from KM.

import Foundation
import CoreGraphics

enum PlayMode: String {
    case openSpace
    case journey
    case hazardLab
}

struct RunProfile {
    var mode: PlayMode
    var level: Int
    var title: String
    var isEndless: Bool
    var goalKm: CGFloat
    var difficulty: CGFloat
    var types: [String]
    var focusType: String?
    var pairTheme: String?
    var comboTheme: String?
    var encounterCount: Int
    var focusChance: CGFloat
    var simpleChance: CGFloat
    var allowAdjacentSetPieces: Bool
    var obstaclesFromKm: CGFloat
    var collectiblesFromKm: CGFloat
    var shieldsFromKm: CGFloat
    var wallBoostsFromKm: CGFloat
    var density0: CGFloat
    var density1: CGFloat
    var minGapFrac: CGFloat
    var gapSpread: CGFloat
    var speedMultiplier: CGFloat
    var maxOnScreen: Int
    var baseCluster0: Int
    var baseCluster1: Int
    var maxCluster: Int
    var maxRowSpawns: Int
    var sparklesTarget: Int
    var smashTarget: Int
    var starSlots: Int
    var introBeats: [IntroBeat]
    var introMessage: String
    var runsMilestones: Bool
    var usesOpenSpaceDensity: Bool
    var usesOpenSpaceCluster: Bool
    var usesKmUnlocks: Bool
    var randomizeFocusEachPick: Bool
    var descriptor: JourneyLevelSpec?

    static let neverKm: CGFloat = 1_000_000_000

    static func openSpace() -> RunProfile {
        RunProfile(
            mode: .openSpace,
            level: 0,
            title: "Open Space",
            isEndless: true,
            goalKm: neverKm,
            difficulty: 0,
            types: GameConfig.Unlocks.table.map(\.type),
            focusType: nil,
            pairTheme: nil,
            comboTheme: nil,
            encounterCount: 0,
            focusChance: 0.5,
            simpleChance: GameConfig.Profile.simpleChance,
            allowAdjacentSetPieces: false,
            obstaclesFromKm: 0,
            collectiblesFromKm: GameConfig.Profile.collectiblesFromScore,
            shieldsFromKm: GameConfig.Profile.shieldsFromScore,
            wallBoostsFromKm: GameConfig.Profile.wallBoostsFromScore,
            density0: GameConfig.Obstacles.scaling.startDensity,
            density1: GameConfig.Obstacles.scaling.maxDensity,
            minGapFrac: 0.25,
            gapSpread: 1.6,
            speedMultiplier: 1.1,
            maxOnScreen: 64,
            baseCluster0: 2,
            baseCluster1: 2,
            maxCluster: 5,
            maxRowSpawns: GameConfig.Profile.maxRowSpawns,
            sparklesTarget: 0,
            smashTarget: 0,
            starSlots: 0,
            introBeats: [],
            introMessage: "",
            runsMilestones: true,
            usesOpenSpaceDensity: true,
            usesOpenSpaceCluster: true,
            usesKmUnlocks: true,
            randomizeFocusEachPick: false,
            descriptor: nil
        )
    }

    static func journey(level raw: Int) -> RunProfile {
        let spec = JourneyConfig.level(raw)
        let neverObstacles = spec.level <= 1
        let neverSparkles = spec.level < JourneyConfig.pointsFromLevel
        let neverShields = spec.level < JourneyConfig.shieldsFromLevel
        let late = spec.level >= 20
        let tLate = (spec.difficulty - 0.72) / 0.28
        return RunProfile(
            mode: .journey,
            level: spec.level,
            title: "Level \(spec.level)",
            isEndless: false,
            goalKm: spec.goalKm,
            difficulty: spec.difficulty,
            types: spec.types,
            focusType: spec.focusType,
            pairTheme: spec.pairTheme,
            comboTheme: spec.comboTheme,
            encounterCount: spec.encounterCount,
            focusChance: late ? 0.32 : 0.5,
            simpleChance: late ? lerp(0.40, 0.26, tLate) : lerp(0.70, 0.42, spec.difficulty),
            allowAdjacentSetPieces: false,
            obstaclesFromKm: neverObstacles ? neverKm : 0,
            collectiblesFromKm: neverSparkles ? neverKm : 0,
            shieldsFromKm: neverShields ? neverKm : 0,
            wallBoostsFromKm: GameConfig.Profile.wallBoostsFromScore,
            density0: 1.15,
            density1: 2.05,
            minGapFrac: late ? lerp(0.18, 0.14, tLate) : lerp(0.30, 0.16, spec.difficulty),
            gapSpread: 1.35,
            speedMultiplier: lerp(0.95, 1.38, spec.difficulty),
            maxOnScreen: late ? 14 : lerpInt(5, 10, spec.difficulty),
            baseCluster0: 1,
            baseCluster1: 4,
            maxCluster: lerpInt(3, 5, spec.difficulty),
            maxRowSpawns: lerpInt(2, 3, spec.difficulty),
            sparklesTarget: spec.sparklesTarget,
            smashTarget: spec.smashTarget,
            starSlots: spec.starSlots,
            introBeats: JourneyConfig.introBeats(for: spec.level),
            introMessage: JourneyConfig.message(for: spec.level),
            runsMilestones: false,
            usesOpenSpaceDensity: false,
            usesOpenSpaceCluster: false,
            usesKmUnlocks: false,
            randomizeFocusEachPick: false,
            descriptor: spec
        )
    }

    static func hazardLab() -> RunProfile {
        RunProfile(
            mode: .hazardLab,
            level: 0,
            title: "Hazard Lab",
            isEndless: false,
            goalKm: HazardLabConfig.goalKm,
            difficulty: HazardLabConfig.difficulty,
            types: HazardLabConfig.types,
            focusType: nil,
            pairTheme: nil,
            comboTheme: nil,
            encounterCount: 0,
            focusChance: 0.55,
            simpleChance: 0.1,
            allowAdjacentSetPieces: true,
            obstaclesFromKm: 0,
            collectiblesFromKm: 0,
            shieldsFromKm: 0,
            wallBoostsFromKm: neverKm,
            density0: 1.2,
            density1: 1.6,
            minGapFrac: lerp(0.28, 0.2, HazardLabConfig.difficulty),
            gapSpread: 1.35,
            speedMultiplier: lerp(0.95, 1.38, HazardLabConfig.difficulty),
            maxOnScreen: lerpInt(5, 8, HazardLabConfig.difficulty),
            baseCluster0: 1,
            baseCluster1: 1,
            maxCluster: 2,
            maxRowSpawns: 2,
            sparklesTarget: 0,
            smashTarget: 0,
            starSlots: 0,
            introBeats: [IntroBeat(text: HazardLabConfig.intro, gapAfterMs: GeneratedJourneyData.defaultBeatGapMs)],
            introMessage: HazardLabConfig.intro,
            runsMilestones: false,
            usesOpenSpaceDensity: false,
            usesOpenSpaceCluster: false,
            usesKmUnlocks: false,
            randomizeFocusEachPick: true,
            descriptor: nil
        )
    }

    func unlockedTypes(scoreKm: CGFloat) -> [String] {
        if usesKmUnlocks {
            return GameConfig.Unlocks.table.compactMap { scoreKm >= $0.score ? $0.type : nil }
        }
        return types
    }

    func density(scoreKm: CGFloat) -> CGFloat {
        if usesOpenSpaceDensity {
            return OpenSpaceBelt.at(scoreKm).density
        }
        return Self.lerp(density0, density1, difficulty)
    }

    func gapRange(height: CGFloat, scoreKm: CGFloat = 0) -> (min: CGFloat, max: CGFloat) {
        if mode == .openSpace {
            let belt = OpenSpaceBelt.at(scoreKm)
            let minGap = height * belt.minGapFrac
            return (minGap, minGap * belt.gapSpread)
        }
        let minGap = height * minGapFrac
        return (minGap, minGap * gapSpread)
    }

    func clusterCount(scoreKm: CGFloat, dens: CGFloat, roll: CGFloat) -> Int {
        let base: Int
        if usesOpenSpaceCluster {
            base = 2 + Int(scoreKm / 8000)
        } else {
            base = Self.lerpInt(baseCluster0, baseCluster1, difficulty)
        }
        let extra = Int(roll * dens)
        return min(maxCluster, max(1, base + extra))
    }

    func advancedBlackHoles(scoreKm: CGFloat) -> Bool {
        switch mode {
        case .openSpace: return scoreKm > 1000
        case .journey: return difficulty >= 0.7
        case .hazardLab: return true
        }
    }

    var isLateJourney: Bool { mode == .journey && level >= 20 }

    func usesPairedBelt(scoreKm: CGFloat) -> Bool {
        if mode == .journey { return level >= 6 }
        if mode == .openSpace { return scoreKm >= GeneratedJourneyData.openSpacePairedFromKm }
        return false
    }

    func livePairTheme(scoreKm: CGFloat) -> String? {
        if mode == .openSpace { return OpenSpaceWeather.at(scoreKm).pair }
        return pairTheme
    }

    func liveComboTheme(scoreKm: CGFloat) -> String? {
        if mode == .openSpace { return OpenSpaceWeather.at(scoreKm).combo }
        return comboTheme
    }

    func liveSimpleChance(scoreKm: CGFloat) -> CGFloat {
        if mode == .openSpace { return OpenSpaceBelt.at(scoreKm).simpleChance }
        return simpleChance
    }

    func rollRowSpawnCount(_ rng: inout UInt64, scoreKm: CGFloat = 0) -> Int {
        let maxSpawns = max(1, maxRowSpawns)
        if maxSpawns <= 1 { return 1 }
        if mode == .openSpace {
            let belt = OpenSpaceBelt.at(scoreKm)
            let r = CombatSimulator.rand01(&rng)
            if r < belt.rowOne { return 1 }
            if r < belt.rowTwo { return min(2, maxSpawns) }
            return min(3, maxSpawns)
        }
        if isLateJourney {
            let r = CombatSimulator.rand01(&rng)
            if r < 0.35 { return 1 }
            if r < 0.80 { return min(2, maxSpawns) }
            return min(3, maxSpawns)
        }
        var spawnCount = 1
        if maxSpawns >= 2, CombatSimulator.rand01(&rng) >= 0.7 { spawnCount = 2 }
        if maxSpawns >= 3, CombatSimulator.rand01(&rng) >= 0.9 { spawnCount = 3 }
        return min(spawnCount, maxSpawns)
    }

    func liveFocusType(_ rng: inout UInt64, scoreKm: CGFloat = 0) -> String? {
        if mode == .openSpace { return OpenSpaceWeather.at(scoreKm).focus }
        if randomizeFocusEachPick, !HazardLabConfig.types.isEmpty {
            rng = rng &* 6364136223846793005 &+ 1
            let idx = Int((rng >> 33) % UInt64(HazardLabConfig.types.count))
            return HazardLabConfig.types[idx]
        }
        return focusType
    }

    private static func lerp(_ a: CGFloat, _ b: CGFloat, _ t: CGFloat) -> CGFloat {
        a + (b - a) * t
    }

    private static func lerpInt(_ a: Int, _ b: Int, _ t: CGFloat) -> Int {
        Int(round(lerp(CGFloat(a), CGFloat(b), t)))
    }
}

enum PlayLaunch: Equatable {
    case openSpace
    case journey(Int)
    case hazardLab

    var profile: RunProfile {
        switch self {
        case .openSpace: return .openSpace()
        case .journey(let level): return .journey(level: level)
        case .hazardLab: return .hazardLab()
        }
    }

    var isLevelRun: Bool {
        switch self {
        case .openSpace: return false
        case .journey, .hazardLab: return true
        }
    }
}
